### ISSUE TITLE
learn.summary() fails on MixedInputModel learner

### ISSUE THREAD

**Issue by stas00** [fastai/fastai#649]
learn.summary() works in dl1/lesson7-cifar10.ipynb, but not in dl1/lesson3-rossman.ipynb

```
m = md.get_learner(emb_szs, n_cont=(len(df.columns)-len(cat_vars)),
                   emb_drop=0.04, out_sz=1, szs=[1000,500], drops=[0.001,0.01], y_range=y_range)
m.summary()
```

to reproduce you need to add `m.summary` to the current notebook.

```
AttributeError                            Traceback (most recent call last)
<ipython-input-87-6c8f0cc6b7e3> in <module>()
      1 m = md.get_learner(emb_szs, n_cont=(len(df.columns)-len(cat_vars)),
      2                    emb_drop=0.04, out_sz=1, szs=[1000,500], drops=[0.001,0.01], y_range=y_range)
----> 3 m.summary()
      4 lr = 1e-3

/mnt/disc1/fast.ai/fastai/courses/dl1/fastai/column_data.py in summary(self)
    149 
    150     def summary(self):
--> 151         x = [torch.ones(3, self.data.trn_ds.cats.shape[1], dtype=torch.int64), torch.rand(3, self.data.trn_ds.conts.shape[1])]
    152         return model_summary(self.model, x)
    153 

AttributeError: module 'torch' has no attribute 'int64'
```

We are still on pytorch 0.3.1, AFAIK, while `torch.int64` is from 0.4.0 API.

Switching it to `long()` fixes it up (in 3 places):

```
--- a/fastai/column_data.py
+++ b/fastai/column_data.py
@@ -142,7 +142,7 @@ class StructuredLearner(Learner):
     def _get_crit(self, data): return F.mse_loss if data.is_reg else F.binary_cross_entropy if data.is_multi else F.nll_loss

     def summary(self):
-        x = [torch.ones(3, self.data.trn_ds.cats.shape[1], dtype=torch.int64), torch.rand(3, self.data.trn_ds.conts.shape[1])]
+        x = [torch.ones(3, self.data.trn_ds.cats.shape[1]).long(), torch.rand(3, self.data.trn_ds.conts.shape[1])]
         return model_summary(self.model, x)

@@ -216,7 +216,7 @@ class CollabFilterLearner(Learner):

     def _get_crit(self, data): return F.mse_loss

-    def summary(self): return model_summary(self.model, [torch.ones(3, dtype=torch.int64), torch.ones(3, dtype=torch.int64)])
+    def summary(self): return model_summary(self.model, [torch.ones(3).long(), torch.ones(3).long()])

```

**Comment by sgugger** [comment]
This CUDA error message often comes from a bad index (though it might be something else). Basically to build the summary, we have to get an input through it and analyze the shapes of the different tensors that come through each steps. Here, the fake input built seems to have the wrong shape?
I'll try to go through it when I have time.

**Comment by stas00** [comment]
I did some more debug and found out that some of the embeddings were getting cardinality of 1, and it must be at least 2.

```
 [(1116, 50),
  (8, 4),
  (4, 2),
  (13, 7),
  (32, 16),
  (3, 2),
  (26, 13),
  (27, 14),
  (5, 3),
  (4, 2),
  (4, 2),
  (24, 12),
  (9, 5),
  (13, 7),
  (53, 27),
  (22, 11),
  (1, 1),
  (1, 1),
  (1, 1),
  (1, 1),
  (1, 1),
  (1, 1)]
```

I'm still investigating how that happened, but meanwhile I added some defensive validation of embeddings, if it's acceptable:

```
--- a/fastai/column_data.py
+++ b/fastai/column_data.py
@@ -89,6 +89,7 @@ class MixedInputModel(nn.Module):
     def __init__(self, emb_szs, n_cont, emb_drop, out_sz, szs, drops,
                  y_range=None, use_bn=False, is_reg=True, is_multi=False):
         super().__init__()
+        for i,(c,s) in enumerate(emb_szs): assert c > 1, f"cardinality must be >=2, got emb_szs[{i}]: ({c},{s})"
         self.embs = nn.ModuleList([nn.Embedding(c, s) for c,s in emb_szs])
         for emb in self.embs: emb_init(emb)
         n_emb = sum(e.embedding_dim for e in self.embs)
```

now the invalid embeddings above get rejected with:

`AssertionError: cardinality must be >=2, got emb_szs[16]: (1,1)`

**Comment by stas00** [comment]
I found the culprit, the notebook code had this marvel:

```
df = train[columns]
df = test[columns]
```

so that the data was missed, and resulting in categories with just `NaN`s in them.

Deleting the 2nd line fixes things up, but I have a question here - what if some values of the categories appear only in the test set and not in the train set? I think it's much safer to combine the two and ensure we get them all. i.e.:

`df = train[columns].append(test[columns])`

I tested that the combination of the two works.
